### PR TITLE
Privacy mode feature

### PR DESCRIPTION
**Pull request by banklesss**
https://user-images.githubusercontent.com/105349292/178338449-ef51a5b3-695c-42b9-a06a-ed695779b2cd.mov

### CODE REVIEW

File: src/Settings/PrivacyMode/PrivacyModeContext.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+  if (!query.data) throw new Error('Invalid state')
+
+  return query.data
+}
+
+const useSavePrivacyMode = ({onSuccess, ...options}: UseMutationOptions<void, Error, PrivacyModeStatus> = {}) => {
+  const queryClient = useQueryClient()
+
+  const mutation = useMutation({
```
Comment (wolverineks) [inline src/Settings/PrivacyMode/PrivacyModeContext.tsx]: useMutationWithInvalidations

File: src/components/HideableText/HideableText.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import {usePrivacyModeContext} from '../../Settings/PrivacyMode/PrivacyModeContext'
+import {Text} from '..'
+
+type Props = {
+  children: string
+}
+
+export const HideableText = ({children}: Props) => {
+  const {privaceModeStatus} = usePrivacyModeContext()
+  return <Text>{privaceModeStatus === 'SHOWN' ? children : children.replaceAll(/./g, '*')}</Text>
```
Comment (jemubm) [inline src/components/HideableText/HideableText.tsx:12]: The fact of hiding all the characters or ignoring the spaces and line breaks, is something that will have to be consulted with the product/design, right?

File: src/components/HideableText/HideableText.tsx [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import {usePrivacyModeContext} from '../../Settings/PrivacyMode/PrivacyModeContext'
+import {Text} from '..'
+
+type Props = {
+  children: string
+}
+
+export const HideableText = ({children, ...props}: Props) => {
+  const {privaceModeStatus} = usePrivacyModeContext()
+  return <Text {...props}>{privaceModeStatus === 'SHOWN' ? children : children.replaceAll(/./g, '\u25CF')}</Text>
```
Comment (banklesss) [inline src/components/HideableText/HideableText.tsx]: Hiding all the characters or ignoring the spaces and line breaks, is something that will have to be consulted with the product/design, right?

File: src/Settings/PrivacyMode/PrivacyModeContext.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
+import AsyncStorage from '@react-native-async-storage/async-storage'
+import React from 'react'
+import {useMutation, UseMutationOptions, useQuery, useQueryClient} from 'react-query'
+
+import {PrivacyModeStatus} from '../../types'
```
Comment (wolverineks) [inline src/Settings/PrivacyMode/PrivacyModeContext.tsx]: `PrivacyModeStatus` -> `PrivacyMode`
and the same for all the others

File: src/Settings/PrivacyMode/PrivacyModeContext.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+export const usePrivacyModeContext = () => React.useContext(PrivacyModeContext) || missingProvider()
+
+const missingProvider = () => {
+  throw new Error('PrivacyModeProvider is missing')
+}
+
+export const usePrivacyMode = () => {
+  const query = useQuery<PrivacyModeStatus, Error>({
+    initialData: defaultPrivacyModeStatus,
```
Comment (wolverineks) [inline src/Settings/PrivacyMode/PrivacyModeContext.tsx]: no initial data

File: src/components/HideableText/HideableText.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import {storiesOf} from '@storybook/react-native'
+import React from 'react'
+import {StyleSheet} from 'react-native'
+import {SafeAreaView} from 'react-native-safe-area-context'
+
+import {usePrivacyModeContext} from '../../Settings/PrivacyMode/PrivacyModeContext'
+import {Button, Spacer, Text} from '..'
+import {HideableText} from './HideableText'
+
+const TEXT = 'Lorem ipsum dolor sit amet'
```
Comment (wolverineks) [inline src/components/HideableText/HideableText.stories.tsx]: move inline

File: src/components/HideableText/HideableText.tsx (hunk 3 above)
Comment (wolverineks) [reply src/components/HideableText/HideableText.tsx]: hide everything for now

File: src/Settings/PrivacyMode/PrivacyModeContext.tsx [hunk 7]
```diff
@@ -0,0 +1,73 @@
+import AsyncStorage from '@react-native-async-storage/async-storage'
```
Comment (wolverineks) [inline src/Settings/PrivacyMode/PrivacyModeContext.tsx]: rename file to privacyMode.ts
move everything in here

**Review by wolverineks — CHANGES_REQUESTED** [review]

File: src/index.tsx [hunk 8]
```diff
@@ -70,7 +71,9 @@ const AppWithProviders = () => {
             <ThemeProvider>
               <LanguageProvider>
                 <CurrencyProvider>
-                  <App />
+                  <PrivacyModeProvider>
```
Comment (stackchain) [inline src/index.tsx]: Suggestion: Move all app settings into an `AppSettingsProviderWrapper`.

Comment (wolverineks) [reply src/index.tsx]: lets do this in a separate pr

File: src/Settings/PrivacyMode/PrivacyModeContext.tsx (hunk 1 above)
Comment (banklesss) [reply src/Settings/PrivacyMode/PrivacyModeContext.tsx]: Done

File: src/Settings/PrivacyMode/PrivacyModeContext.tsx (hunk 4 above)
Comment (banklesss) [reply src/Settings/PrivacyMode/PrivacyModeContext.tsx]: Done

File: src/Settings/PrivacyMode/PrivacyModeContext.tsx (hunk 5 above)
Comment (banklesss) [reply src/Settings/PrivacyMode/PrivacyModeContext.tsx]: Done

File: src/components/HideableText/HideableText.stories.tsx (hunk 6 above)
Comment (banklesss) [reply src/components/HideableText/HideableText.stories.tsx]: Done

File: src/Settings/PrivacyMode/PrivacyModeContext.tsx (hunk 7 above)
Comment (banklesss) [reply src/Settings/PrivacyMode/PrivacyModeContext.tsx]: Done

File: src/Settings/PrivacyMode/PrivacyModeContext.tsx (hunk 4 above)
Comment (wolverineks) [reply src/Settings/PrivacyMode/PrivacyModeContext.tsx]: also, using react query we dont need a context at all

File: src/Settings/PrivacyMode/PrivacyMode.tsx [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    mutationFn: async (privacyMode) => AsyncStorage.setItem('/appSettings/privacyMode', JSON.stringify(privacyMode)),
+    invalidateQueries: [['privacyMode']],
+    ...options,
+  })
+
+  return mutation.mutate
+}
+
+type PrivacyMode = 'SHOWN' | 'HIDDEN'
+const defaultPrivacyMode = 'SHOWN' as PrivacyMode
```
Comment (stackchain) [inline src/Settings/PrivacyMode/PrivacyMode.tsx]: Casting is not safe.

```suggestion
const defaultPrivacyMode: PrivacyMode = 'SHOWN'
```

**Review by stackchain — CHANGES_REQUESTED** [review]

**Pull request merged**